index/miner/lotusidx/onchain.go

```diff
@@ -0,0 +1,127 @@
+package lotusidx
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/lotus/api/apistruct"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/textileio/powergate/v2/index/miner"
+)
+
+func (mi *Index) updateOnChainIndex(ctx context.Context) error {
+	start := time.Now()
+	log.Info("updating on-chain index...")
+	defer log.Info("on-chain index updated")
+
+	client, cls, err := mi.cb(context.Background())
+	if err != nil {
+		return fmt.Errorf("creating lotus client: %s", err)
+	}
+	defer cls()
+
+	listStart := time.Now()
+	addrs, err := client.StateListMiners(ctx, types.EmptyTSK)
+	if err != nil {
+		return fmt.Errorf("calling state-list-miners: %s", err)
+	}
+	log.Infof("listing all on-chain miners took %.2f seconds", time.Since(listStart).Seconds())
+
+	newIndex := miner.ChainIndex{
+		Miners: map[string]miner.OnChainMinerData{},
+	}
+	if err := mi.updateForAddrs(ctx, client, &newIndex, addrs); err != nil {
+		return fmt.Errorf("updating for addresses: %s", err)
+	}
+	chainHead, err := client.ChainHead(ctx)
+	if err != nil {
+		return fmt.Errorf("getting chain head: %s", err)
+	}
+	newIndex.LastUpdated = int64(chainHead.Height())
+
+	mi.lock.Lock()
+	mi.index.OnChain = newIndex
+	mi.lock.Unlock()
+
+	if err := mi.store.SaveOnChain(ctx, newIndex); err != nil {
```

Comment:
Some code moving here, and removing som old style of saving on-chain data which had problems to saving when using MongoDB as the backing store for go-datastore.

Before, we simply saved the whole miner index as a single JSON in go-datastore. But MongoDB has a 16MB limit on document sizes, so Powergate failed to saved the index. This PR created a new `store` for the miner-index which saves Miner Index information per-miner, so avoiding big documents.